app/src/main/java/com/pvp/app/ui/screen/friends/FriendsViewModel.kt

```diff
@@ -0,0 +1,142 @@
+            user.collect { user ->
+                user?.email?.let { email ->
+                    friendService.createFriendObject(email)
+                }
+            }
+        }
+    }
+
+    fun createFriendObject(email: String) {
+        viewModelScope.launch {
+            friendService.createFriendObject(email)
+        }
+    }
+
+    fun addFriend(friendEmail: String){
+        viewModelScope.launch {
+            if (friendEmail.isEmpty()) {
+                toastMessage.value = "Please enter an email"
+                _isRequestSent.value = false
+                return@launch
+            }
+
+            val userEmail = user.value?.email ?: return@launch
+            val friendObject = friendService.get(userEmail).firstOrNull() ?: return@launch
+            val friendUser = userService.get(friendEmail).firstOrNull()
+
+            if (friendUser == null) {
+                toastMessage.value = "User with email $friendEmail does not exist"
+                _isRequestSent.value = false
+                return@launch
+            }
+
+            friendService.createFriendObject(friendEmail)
+
+            val toastMessageValue = friendService.addFriend(
+                friendObject,
+                userEmail,
+                friendEmail
+            )
+            toastMessage.value = toastMessageValue
+            _isRequestSent.value = toastMessageValue == "Friend request sent!"
+        }
+    }
+
+    fun acceptFriendRequest(friendEmail: String) {
+        viewModelScope.launch {
+            val userEmail = user.value?.email ?: return@launch
+            val friendObject = friendService.get(userEmail).firstOrNull() ?: return@launch
+
+            toastMessage.value = friendService.acceptFriendRequest(
+                friendObject,
+                userEmail,
+                friendEmail
+            )
+        }
+    }
+
+    fun denyFriendRequest(friendEmail: String) {
+        viewModelScope.launch {
+            val userEmail = user.value?.email ?: return@launch
+            val friendObject = friendService.get(userEmail).firstOrNull() ?: return@launch
+
+            toastMessage.value = friendService.denyFriendRequest(
+                friendObject,
+                userEmail,
+                friendEmail
+            )
+        }
+    }
+
+    fun cancelSentRequest(friendEmail: String) {
+        viewModelScope.launch {
+            val userEmail = user.value?.email ?: return@launch
+            val friendObject = friendService.get(userEmail).firstOrNull() ?: return@launch
+
+            toastMessage.value = friendService.cancelSentRequest(
+                friendObject,
+                userEmail,
+                friendEmail
+            )
+        }
+    }
+
+    fun getFriendAvatar(friendEmail: String): ImageBitmap {
+        var avatar: ImageBitmap? = null
+
+        viewModelScope.launch {
+            avatar = userService.resolveAvatar(friendEmail)
+        }
+
+        return avatar ?: ImageBitmap(0, 0)
```

Comment:
ImageBitmap of 0px width/height is going to throw an error within certain chained constructor, I believe.